test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Internal/ValidationStackTest.cs

```diff
@@ -0,0 +1,132 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.ModelBinding.Internal;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.Core.Test.ModelBinding.Internal
+{
+    public class ValidationStackTest
+    {
+        [Theory]
+        [InlineData(0)]
+        [InlineData(5)]
+        [InlineData(ValidationStack.CutOff + 1)]
+        public void ValidationStack_PushFalseWhenAlreadyIn(int preloadCount)
+        {
+            // Arrange
+            var validationStack = new TestValidationStack();
+            var model = "This is a value";
+
+            PreLoad(preloadCount, validationStack);
+
+            // Act & Assert
+            Assert.True(validationStack.Push(model));
+            Assert.False(validationStack.Push(model));
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(5)]
+        [InlineData(ValidationStack.CutOff + 1)]
+        public void ValidationStack_PushAndPop(int preloadCount)
+        {
+            // Arrange
+            var validationStack = new TestValidationStack();
+            var model = "This is a value";
+
+            PreLoad(preloadCount, validationStack);
+
+            // Act
+            validationStack.Push(model);
+            validationStack.Pop(model);
+
+            // Assert
+            Assert.False(validationStack.Contains(model));
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(5)]
+        [InlineData(ValidationStack.CutOff + 1)]
+        public void ValidationStack_PopNull(int preloadCount)
```

Comment:
`Pop_DoesNotThrowIfValueIsNull`. Additionally, does this require the stack to be preloaded? Does this work when the stack is empty?